### PR TITLE
[Fix #1772] Autocomplete fails when nREPL not connected

### LINKED ISSUE #1772: Autocomplete fails when nREPL not connected
## Expected behavior

Can auto complete typing as well even if nREPL not connected.

## Actual behavior

When nREPL not connected, it will show below error message when typing:

```
Error running timer `ac-update-greedy': (wrong-type-argument stringp nil)
Error running timer `ac-show-menu': (wrong-type-argument stringp nil)
```

## Steps to reproduce the problem

1. Launch Emacs
2. Switch to a temporary buffer by `C-x b tmp-buffer`
3. `M-x clojure-mode`
4. `M-x cider-mode`
5. Type something, like `hello world`, to trigger `auto-complete`

## Related code

cider-client.el:cider-current-session

```
(defun cider-current-session ()
  "The REPL session to use for this buffer."
  (with-current-buffer (cider-current-connection)
    nrepl-session))
```

`(cider-current-connection)` will return nil if nREPL not connected.

### CIDER version information

```
;; CIDER 0.12.0 (Seattle), nREPL 0.2.12
;; Clojure 1.8.0, Java 1.8.0_45
```

### Emacs version

GNU Emacs 24.5.1 (x86_64-apple-darwin14.1.1, NS apple-appkit-1345.38)

### Operating system

OS X 10.11 "El Capitan"

### PR DESCRIPTION
**Pull request by redraiment**
Before submitting the PR make sure the following things have been done (and denote this
by checking the relevant checkboxes):

- [x] The commits are consistent with our [contribution guidelines][1]
- [ ] You've added tests (if possible) to cover your change(s)
- [x] All tests are passing (`make test`)
- [x] The new code is not generating bytecode or `M-x checkdoc` warnings
- [ ] You've updated the changelog (if adding/changing user-visible functionality)
- [ ] You've updated the readme (if adding/changing user-visible functionality)
- [ ] You've updated the refcard (if you made changes to the commands listed there)

Thanks!

[1]: https://github.com/clojure-emacs/cider/blob/master/.github/CONTRIBUTING.md

### CODE REVIEW

File: cider-client.el [hunk 1]
```diff
@@ -859,13 +859,14 @@ Optional arguments include SEARCH-NS, DOCS-P, PRIVATES-P, CASE-SENSITIVE-P."
...
   "Return a list of completions for STR using nREPL's \"complete\" op.
 CONTEXT represents a completion context for compliment."
-  (when-let ((dict (thread-first (list "op" "complete"
-                                       "session" (cider-current-session)
-                                       "ns" (cider-current-ns)
-                                       "symbol" str
-                                       "context" context)
-                     (cider-nrepl-send-sync-request nil 'abort-on-input))))
-    (nrepl-dict-get dict "completions")))
+  (if (cider-current-connection)
```
Comment (bbatsov) [inline cider-client.el]: `(when (cider-connected-p)` would be better.

Comment (redraiment) [inline cider-client.el]: Changed. Thanks for suggestion.

**Comment by Malabarba** [comment]
Thanks for the effort. This seems like the wrong way to fix this.
None of our `-request:*` functions check for connection. 

I think the proper fix would be to avoid calling this function in the first place when there's no connection.

**Comment by bbatsov** [comment]
> I think the proper fix would be to avoid calling this function in the first place when there's no connection.
>  Show all checks

I was about to mention exactly this in the morning, but got distracted and forgot about it. You're completely right.

**Comment by bbatsov** [comment]
@redraiment Please, move the check higher up the call stack, as suggested by @Malabarba.

**Comment by redraiment** [comment]
Sure. Thanks! I'm going to do it.

**Comment by bbatsov** [comment]
OK, let me know when you do this.

**Comment by redraiment** [comment]
After research, it would be better to fixed in ac-cider project rather than cider.

**Pull request closed without merging**